src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3621,13 +3621,61 @@ private TypeWithState InferResultNullability(BinaryOperatorKind operatorKind, Me
         protected override void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOperator> stack)
         {
             var binary = stack.Pop();
-
             var (leftOperand, leftConversion) = RemoveConversion(binary.Left, includeExplicitConversions: false);
-            Visit(leftOperand);
+
+            // Only the leftmost operator of a left-associative binary operator chain can learn from a conditional access on the left
+            // For simplicity, we just special case it here.
+            // For example, `a?.b(out x) == true` has a conditional access on the left of the operator,
+            // but `expr == a?.b(out x) == true` has a conditional access on the right of the operator
+            if (VisitPossibleConditionalAccess(leftOperand, out var conditionalStateWhenNotNull)
+                && CanPropagateStateWhenNotNull(leftConversion)
+                && isEqualsOrNotEquals(binary))
+            {
+                Debug.Assert(!IsConditionalState);
+                var leftType = ResultType;
+                var (rightOperand, rightConversion) = RemoveConversion(binary.Right, includeExplicitConversions: false);
+
+                // Consider the following two scenarios:
+                // `a?.b(x = new object()) == c.d(x = null)` // `x` is maybe-null after expression
+                // `a?.b(x = null) == c.d(x = new object())` // `x` is not-null after expression
+
+                // In order to properly integrate state changes from the RHS into the LHS's "state when not null", we can't simply Meet or Join the state.
```

Comment:
nit: Consider refreshing the comment in light of our discussion today. Maybe something like:

// If the comparison is amenable:
// We take the when-not-null state from visiting the LHS and use that as starting point to visit the RHS ignoring diagnostics. The resulting state is the when-true state resulting from the comparison.
// We take the state (ie. worst case state) from visiting the LHS and use that as starting point to visit the RHS for diagnostics and recording nullabilities for public APIs. The resulting state is the when-false state resulting from the comparison.

Reply:
> The resulting state is the when-true state

Whether the resulting state is for when-true or when-false of the containing comparison depends on the operator and the "other" operand in use, roughly according to the following

| Operator | Other operand | StateWhenNotNull moves to |
|----------|---------------|---------------------------|
| ==       | non-null      | StateWhenTrue             |
| ==       | constant null | StateWhenFalse            |
| !=       | non-null      | StateWhenFalse            |
| !=       | constant null | StateWhenTrue             |

Reply:
I'm fine if the comment only explains `==`. The swapping for `!=` was not hard to follow in my opinion.

Reply:
Have updated the comment.